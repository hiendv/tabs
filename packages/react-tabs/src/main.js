export { default as Tab } from './Tab.jsx'
export { default as Tabs } from './Tabs.jsx'
export * from '@hiendv/tabs'
