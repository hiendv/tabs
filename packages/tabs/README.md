# @hiendv/tabs
Utilities & themes for Tabs components

## Installation & Usage
### CDN
```html
<script src="https://cdn.jsdelivr.net/npm/@hiendv/tabs@latest/dist/iife.js" crossorigin="anonymous"></script>
```
```js
console.log(tabs)
/*
{
  styleDark: {tabs: "dark-module_tabs__2D2d-", items: "dark-module_items__vLUYO", item: "dark-module_item__2t_72", item--active: "dark-module_item--active__2BwxE", panel: "dark-module_panel__3HSBq"},
  styleDefault: {tabs: "default-module_tabs__3f18F", items: "default-module_items__4cVRy", item: "default-module_item__2w-4Z", item--active: "default-module_item--active__3kx5I", panel: "default-module_panel__1onVq"}
}
*/
```

### NPM
The package is mostly consumed by internal Vue & React components. Everything is re-exported.
```js
// Vue
import { styleDefault, styleDark } from '@hiendv/vue-tabs'
console.log(styleDefault, styleDark)
```
