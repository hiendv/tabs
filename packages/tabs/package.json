{
  "name": "@hiendv/tabs",
  "version": "0.1.2",
  "module": "dist/es.js",
  "main": "dist/es.js",
  "author": "Hien Dao <hien.dv.neo@gmail.com>",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@hiendv/bem-sass": "^0.1.0"
  },
  "files": [
    "dist",
    "styles",
    "README.md"
  ]
}
