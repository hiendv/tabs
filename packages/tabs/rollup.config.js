import path from 'path'
import postcss from 'rollup-plugin-postcss'

const reslv = p => {
  return path.resolve(__dirname, p)
}

export default [
  {
    input: reslv('src/index.js'),
    output: {
      format: 'es',
      file: reslv('dist/es.js')
    },
    external: id => /\.scss$/.test(id)
  },
  {
    input: reslv('src/index.js'),
    output: {
      name: 'tabs',
      format: 'iife',
      file: reslv('dist/iife.js')
    },
    plugins: [postcss()]
  }
]
