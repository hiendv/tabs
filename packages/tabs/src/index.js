export { default as themeDefault } from '../themes/default.module.scss'
export { default as themeDark } from '../themes/dark.module.scss'
export { setHash } from './location'
