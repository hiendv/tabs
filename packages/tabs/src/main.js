export { default as styleDefault } from '../styles/default.module.scss'
export { default as styleDark } from '../styles/dark.module.scss'
