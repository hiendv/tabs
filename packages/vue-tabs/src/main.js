export { default as Tab } from './Tab'
export { default as Tabs } from './Tabs.vue'
export * from '@hiendv/tabs'
