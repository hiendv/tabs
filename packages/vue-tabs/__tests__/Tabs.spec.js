import { mount, createLocalVue } from '@vue/test-utils'
import { setHash } from '@hiendv/tabs'
import VueRouter from 'vue-router'
import { alert, zap } from 'octicons-vue'
import { Tabs, Tab } from '../src/main.js'

let localVue = null
const transitionStub = () => ({
  render: function (h) {
    return this.$options._renderChildren
  }
})

beforeAll(() => {
  localVue = createLocalVue()
  localVue.component('tab', Tab)
  localVue.component('transition', transitionStub())
})

afterAll(() => {
  localVue = null
})

describe('Tabs', () => {
  it('renders', async () => {
    const wrapper = mount(Tabs, {
      slots: {
        default: `
          <tab title="One"><div>One Content</div></tab>
          <tab title="Two"><div>Two Content</div></tab>
        `
      },
      localVue
    })

    expect(wrapper.html()).toMatchSnapshot()

    wrapper.findAll('.item').at(1).trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.emitted()['update:show']).toStrictEqual([[1]])
  })

  it('renders without slots', () => {
    const wrapper = mount(Tabs, { stubs: { transition: transitionStub() } })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders with custom theme', () => {
    const wrapper = mount(Tabs, {
      propsData: {
        theme: {
          tabs: 'one',
          items: 'two',
          item: 'three',
          'item--active': 'three--active',
          panel: 'four'
        }
      },
      slots: {
        default: `
          <tab title="One"><div>One Content</div></tab>
          <tab title="Two"><div>Two Content</div></tab>
        `
      },
      localVue
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders with panel custom bindings', () => {
    const wrapper = mount(Tabs, {
      slots: {
        default: `
          <tab title="One" style="background-color: #f5f5f5" class="foobar"><div>One Content</div></tab>
          <tab title="Two"><div>Two Content</div></tab>
        `
      },
      localVue
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders with custom navigation', async () => {
    const wrapper = mount(Tabs, {
      propsData: {
        show: 1
      },
      slots: {
        default: `
          <tab title="One"><div>One Content</div></tab>
          <tab title="Two"><div>Two Content</div></tab>
        `
      },
      scopedSlots: {
        nav (props) {
          return this.$createElement('div', props.items.map((item, index) => {
            return this.$createElement('a', {
              key: index,
              on: {
                click: e => {
                  e.preventDefault()
                  wrapper.setProps({ show: index })
                }
              }
            }, item.title)
          }))
        }
      },
      localVue
    })

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.props()).toMatchObject({
      show: 1
    })

    wrapper.find('a').trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.props()).toMatchObject({
      show: 0
    })
  })

  it('renders with hashes', () => {
    setHash('two')

    const wrapper = mount(Tabs, {
      slots: {
        default: `
          <tab title="One" hash="one"><div>One Content</div></tab>
          <tab title="Two" hash="two"><div>Two Content</div></tab>
        `
      },
      localVue
    })

    expect(wrapper.html()).toMatchSnapshot()
    wrapper.find('a').trigger('click')
    expect(window.location.hash).toEqual('#one')
  })

  it('renders with icons', () => {
    const wrapper = mount(Tabs, {
      mocks: {
        alert,
        zap
      },
      slots: {
        default: `
          <tab title="One" :icon="alert"><div>One Content</div></tab>
          <tab title="Two" :icon="zap"><div>Two Content</div></tab>
        `
      },
      localVue
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders with hashes from vue-router', () => {
    setHash('')
    localVue.use(VueRouter)
    const router = new VueRouter()

    const wrapper = mount(Tabs, {
      slots: {
        default: `
          <tab title="One" hash="one"><div>One Content</div></tab>
          <tab title="Two" hash="two"><div>Two Content</div></tab>
          <tab title="Three"><div>Three Content</div></tab>
          <tab title="Four"><div>Four Content</div></tab>
        `
      },
      localVue,
      router
    })

    expect(window.location.toString()).toEqual('http://localhost/#/')
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.find('a').trigger('click')
    expect(window.location.toString()).toEqual('http://localhost/#/#one')

    wrapper.findAll('.item').at(1).trigger('click')
    expect(window.location.toString()).toEqual('http://localhost/#/#two')

    wrapper.findAll('.item').at(2).trigger('click')
    expect(window.location.toString()).toEqual('http://localhost/#/')

    wrapper.findAll('.item').at(3).trigger('click')
    expect(window.location.toString()).toEqual('http://localhost/#/')
  })

  it('renders <tab> without slots', async () => {
    const wrapper = mount(Tabs, {
      slots: {
        default: `
          <tab></tab>
          <tab title="Two"></tab>
        `
      },
      localVue
    })

    expect(wrapper.html()).toMatchSnapshot()

    wrapper.findAll('.item').at(1).trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.emitted()['update:show']).toStrictEqual([[1]])
  })

  it('renders ghost <tab>', () => {
    const wrapper = mount(Tabs, {
      slots: {
        default: `
          <tab title="One"><div>One Content</div></tab>
          <tab title="Two"><div>Two Content</div></tab>
          <tab title="Ghost" ghost></tab>
        `
      },
      localVue
    })

    expect(wrapper.html()).toMatchSnapshot()

    wrapper.findAll('.item').at(2).trigger('click')
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.emitted()['update:show']).toBeFalsy()
  })

  it('renders end <tab>', async () => {
    const wrapper = mount(Tabs, {
      slots: {
        default: `
          <tab title="One"><div>One Content</div></tab>
          <tab title="Two" end><div>Two Content</div></tab>
        `
      },
      localVue
    })

    expect(wrapper.html()).toMatchSnapshot()

    wrapper.findAll('.item').at(1).trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.emitted()['update:show']).toStrictEqual([[1]])
  })
})
