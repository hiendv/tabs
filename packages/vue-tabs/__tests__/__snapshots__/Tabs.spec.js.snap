// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs renders <tab> without slots 1`] = `
<div class="tabs">
  <nav class="items"><a href="#" class="item item--active">
      <!----> Title</a><a href="#" class="item">
      <!----> Two</a></nav>
  <div class="panel" tag="div" name="slide-down" mode="out-in"></div>
</div>
`;

exports[`Tabs renders <tab> without slots 2`] = `
<div class="tabs">
  <nav class="items"><a href="#" class="item">
      <!----> Title</a><a href="#" class="item item--active">
      <!----> Two</a></nav>
  <div class="panel" tag="div" name="slide-down" mode="out-in"></div>
</div>
`;

exports[`Tabs renders 1`] = `
<div class="tabs">
  <nav class="items"><a href="#" class="item item--active">
      <!----> One</a><a href="#" class="item">
      <!----> Two</a></nav>
  <div class="panel" tag="div" name="slide-down" mode="out-in">
    <div>One Content</div>
  </div>
</div>
`;

exports[`Tabs renders 2`] = `
<div class="tabs">
  <nav class="items"><a href="#" class="item">
      <!----> One</a><a href="#" class="item item--active">
      <!----> Two</a></nav>
  <div class="panel" tag="div" name="slide-down" mode="out-in">
    <div>Two Content</div>
  </div>
</div>
`;

exports[`Tabs renders end <tab> 1`] = `
<div class="tabs">
  <nav class="items"><a href="#" class="item item--active">
      <!----> One</a><a href="#" class="item item--end">
      <!----> Two</a></nav>
  <div class="panel" tag="div" name="slide-down" mode="out-in">
    <div>One Content</div>
  </div>
</div>
`;

exports[`Tabs renders end <tab> 2`] = `
<div class="tabs">
  <nav class="items"><a href="#" class="item">
      <!----> One</a><a href="#" class="item item--active item--end">
      <!----> Two</a></nav>
  <div class="panel" tag="div" name="slide-down" mode="out-in">
    <div>Two Content</div>
  </div>
</div>
`;

exports[`Tabs renders ghost <tab> 1`] = `
<div class="tabs">
  <nav class="items"><a href="#" class="item item--active">
      <!----> One</a><a href="#" class="item">
      <!----> Two</a><a href="#" class="item">
      <!----> Ghost</a></nav>
  <div class="panel" tag="div" name="slide-down" mode="out-in">
    <div>One Content</div>
  </div>
</div>
`;

exports[`Tabs renders ghost <tab> 2`] = `
<div class="tabs">
  <nav class="items"><a href="#" class="item item--active">
      <!----> One</a><a href="#" class="item">
      <!----> Two</a><a href="#" class="item">
      <!----> Ghost</a></nav>
  <div class="panel" tag="div" name="slide-down" mode="out-in">
    <div>One Content</div>
  </div>
</div>
`;

exports[`Tabs renders with custom navigation 1`] = `
<div class="tabs">
  <div><a>One</a><a>Two</a></div>
  <div class="panel" tag="div" name="slide-down" mode="out-in">
    <div>Two Content</div>
  </div>
</div>
`;

exports[`Tabs renders with custom navigation 2`] = `
<div class="tabs">
  <div><a>One</a><a>Two</a></div>
  <div class="panel" tag="div" name="slide-down" mode="out-in">
    <div>One Content</div>
  </div>
</div>
`;

exports[`Tabs renders with custom theme 1`] = `
<div class="one">
  <nav class="two"><a href="#" class="three three--active">
      <!----> One</a><a href="#" class="three">
      <!----> Two</a></nav>
  <div class="four" tag="div" name="slide-down" mode="out-in">
    <div>One Content</div>
  </div>
</div>
`;

exports[`Tabs renders with hashes 1`] = `
<div class="tabs">
  <nav class="items"><a href="#one" class="item">
      <!----> One</a><a href="#two" class="item item--active">
      <!----> Two</a></nav>
  <div class="panel" tag="div" name="slide-down" mode="out-in">
    <div>Two Content</div>
  </div>
</div>
`;

exports[`Tabs renders with hashes from vue-router 1`] = `
<div class="tabs">
  <nav class="items"><a href="#one" class="item item--active">
      <!----> One</a><a href="#two" class="item">
      <!----> Two</a></nav>
  <div class="panel" tag="div" name="slide-down" mode="out-in">
    <div>One Content</div>
  </div>
</div>
`;

exports[`Tabs renders without slots 1`] = `
<div class="tabs">
  <nav class="items"></nav>
  <div class="panel" tag="div" name="slide-down" mode="out-in"></div>
</div>
`;
